### PR TITLE
Callback telemetry QOS

### PR DESCRIPTION
**Pull request by danewalton**
closes #109

### CODE REVIEW

File: source/azure_iot_hub_client.c [hunk 1]
```diff
@@ -833,6 +864,10 @@ AzureIoTHubClientResult_t AzureIoTHubClient_SendTelemetry( AzureIoTHubClient_t *
         }
         else
         {
+            if( *pusTelemetryPacketID != NULL && xQOS == eAzureIoTHubMessageQoS1 )
```
[the remark this answers is not in the document]
Comment (hihigupt) [reply source/azure_iot_hub_client.c:867]: Please check xQ0S first

[the remark this answers is not in the document]
Comment (danewalton) [reply source/azure_iot_hub_client.c:867]: Sorry yes this very likely has a few errors I coded it up to mainly show the flow in comparison. I will update it if we decide to go with it.

File: source/azure_iot_hub_client.c [hunk 2]
```diff
@@ -126,6 +126,29 @@ static void prvMQTTProcessSuback( AzureIoTHubClient_t * pxAzureIoTHubClient,
...
+
+    configASSERT( pxIncomingPacket != NULL );
+    configASSERT( ( azureiotmqttGET_PACKET_TYPE( pxIncomingPacket->ucType ) ) == azureiotmqttPACKET_TYPE_PUBACK );
+
+    AZLogInfo( ( "Puback received for packet id: 0x%08x", usPacketID ) );
+
+    if( pxAzureIoTHubClient->_internal.xTelemetryCallback != NULL )
+    {
+        pxAzureIoTHubClient->_internal.xTelemetryCallback( usPacketID );
+    }
```
Comment (hihigupt) [inline source/azure_iot_hub_client.c:148]: should we add Debug log for tracing? Like entering callback and done with callback

**Comment by hihigupt** [comment]
>         pxHubClientOptions->pucModelID = NULL;

is this redundant if we are doing explicit initizlization of every filed ? ANd compiler we are checking our unit test always complaint about some varables are not init?

---
Refers to: source/azure_iot_hub_client.c:575 in de75a9b. [](commit_id = de75a9b02716e6e6e817b185c83e689f1c877b1b, deletion_comment = False)

**Comment by hihigupt** [comment]
> 

I do not think we should generate identifier if QoS is zero.

---
Refers to: source/azure_iot_hub_client.c:858 in de75a9b. [](commit_id = de75a9b02716e6e6e817b185c83e689f1c877b1b, deletion_comment = False)

File: source/azure_iot_hub_client.c [hunk 3]
```diff
@@ -815,7 +847,7 @@ AzureIoTHubClientResult_t AzureIoTHubClient_SendTelemetry( AzureIoTHubClient_t *
     }
     else
     {
-        xMQTTPublishInfo.xQOS = eAzureIoTMQTTQoS1;
+        xMQTTPublishInfo.xQOS = xQOS;
```
Comment (hihigupt) [inline source/azure_iot_hub_client.c:850]: >xQOS [](http://example.com/codeflow?start=32&length=4)

xQOS is of different type then MqttQoS, should this result in build failure?

File: tests/e2e/device/e2e_device_commands.c [hunk 4]
```diff
@@ -584,13 +598,15 @@ static uint32_t prvE2ETestReportedPropertiesCommandExecute( E2E_TEST_COMMAND_HAN
     else if( AzureIoTHubClient_SendTelemetry( pxAzureIoTHubClient,
                                               xCMD->pulReceivedData,
                                               xCMD->ulReceivedDataLength,
-                                              NULL ) != eAzureIoTHubClientSuccess )
+                                              NULL, eAzureIoTHubMessageQoS1, &usTelemetryPacketID ) != eAzureIoTHubClientSuccess )
```
Comment (hihigupt) [inline tests/e2e/device/e2e_device_commands.c:601]: how about creating send telemetry wrapper with timeout. In that function you can call for ProcessLoop till that timeout and check for packetId returned and received. To record the received packet have a global state that is shared between the callback and this wapper api?.

Using this way you do not need this complex, setting which might failed if process loop is not called and callback was not called.

File: tests/ut/azure_iot_hub_client_ut.c [hunk 5]
```diff
@@ -495,7 +495,10 @@ static void testAzureIoTHubClient_SendTelemetry_InvalidArgFailure( void ** ppvSt
     /* Fail if the hub client is NULL. */
     assert_int_equal( AzureIoTHubClient_SendTelemetry( NULL,
                                                        ucTestTelemetryPayload,
-                                                       sizeof( ucTestTelemetryPayload ) - 1, NULL ),
+                                                       sizeof( ucTestTelemetryPayload ) - 1,
+                                                       NULL,
+                                                       eAzureIoTHubMessageQoS0,
```
Comment (hihigupt) [inline tests/ut/azure_iot_hub_client_ut.c:500]: >eAzureIoTHubMessageQoS0 [](http://example.com/codeflow?start=55&length=23)

please add QoS1 test cases also. In that test you can be creative to verify that MQTT layer is correctly getting the QoS from middleware layer.

**Review by hihigupt — CHANGES_REQUESTED** [review]
:clock1:

**Review by jspaith — APPROVED** [review]
I don't have additional concerns beyond @hihigupt feedback.  Thanks guys.

File: tests/e2e/device/e2e_device_commands.c (hunk 4 above)
Comment (danewalton) [reply tests/e2e/device/e2e_device_commands.c:645]: You're saying you don't want the ultracomplicated pointer pointer method???? :)

updated

File: source/azure_iot_hub_client.c (hunk 3 above)
Comment (danewalton) [reply source/azure_iot_hub_client.c]: I have it casted now and checked in unit tests to make sure the qos aligns. Whether it should have broken the build or not I think is worthwhile to look into the build flags

Comment (hihigupt) [reply source/azure_iot_hub_client.c]: how about 

xQOS == eAzureIoTHubMessageQoS1 ? eAzureIoTMQTTQoS1 : eAzureIoTMQTTQoS0;


even though we own both enum... its better not to do unnecessary casts

File: source/azure_iot_hub_client.c [hunk 6]
```diff
@@ -815,14 +846,16 @@ AzureIoTHubClientResult_t AzureIoTHubClient_SendTelemetry( AzureIoTHubClient_t *
...
+        xMQTTPublishInfo.xQOS = ( AzureIoTMQTTQoS_t ) xQOS;
         xMQTTPublishInfo.pcTopicName = pxAzureIoTHubClient->_internal.pucWorkingBuffer;
         xMQTTPublishInfo.usTopicNameLength = ( uint16_t ) xTelemetryTopicLength;
         xMQTTPublishInfo.pvPayload = ( const void * ) pucTelemetryData;
         xMQTTPublishInfo.xPayloadLength = ulTelemetryDataLength;
 
-        /* Get a unique packet id. */
-        usPublishPacketIdentifier = AzureIoTMQTT_GetPacketId( &( pxAzureIoTHubClient->_internal.xMQTTContext ) );
+        /* Get a unique packet id. Not used if QOS is 0 */
+        usPublishPacketIdentifier = xQOS == eAzureIoTHubMessageQoS1 ?
```
Comment (hihigupt) [inline source/azure_iot_hub_client.c:856]: >usPublishPacketIdentifier [](http://example.com/codeflow?start=8&length=25)

start declaration with intialization =0

if (xQOS == eAzureIoTHubMessageQoS1)
{
usPublishPacketIdentifier = AzureIoTMQTT_GetPacketId( &( pxAzureIoTHubClient->_interna);
}

File: tests/e2e/device/e2e_device_commands.c [hunk 7]
```diff
@@ -71,6 +71,45 @@ static const uint8_t ucStatusOKTelemetry[] = "{\"status\": \"OK\"}";
...
+{
+    AZLogInfo( ( "Puback received for packet id in callback: 0x%08x", usPacketID ) );
+    usReceivedTelemetryPubackID = usPacketID;
+}
+
+/**
+ * Telemetry PUBACK callback
+ *
+ **/
+static AzureIoTHubClientResult_t prvWaitForPuback( AzureIoTHubClient_t * pxAzureIoTHubClient,
```
Comment (hihigupt) [inline tests/e2e/device/e2e_device_commands.c:91]: >AzureIoTHubClientResult_t [](http://example.com/codeflow?start=7&length=25)

lets make it uint32_t

File: tests/e2e/device/e2e_device_task.c [hunk 8]
```diff
@@ -35,6 +39,8 @@ static void vTestEntry( void ** ppvState )
     TlsTransportParams_t xTlsTransportParams = { 0 };
     AzureIoTHubClientOptions_t xHubOptions = { 0 };
     bool xSessionPresent = false;
+    uint16_t * pusArrayOfSentPacketID;
+    uint16_t totalSentPublishes;
```
Comment (hihigupt) [inline tests/e2e/device/e2e_device_task.c:44]: do you need these?

**Review by hihigupt — APPROVED** [review]
:shipit:

**Pull request merged**